File: hash/go/imagehash/internal/dct/dct_test.go
package dct_test

import (
	"encoding/json"
	"math"
	"os"
	"testing"

	"github.com/wmetcalf/rosetta-image-hash/go/imagehash/internal/dct"
	"github.com/wmetcalf/rosetta-image-hash/go/imagehash/internal/testkit"
)

const tol = 1e-9

func TestDCT1DMatchesScipy(t *testing.T) {
	path := testkit.PathFromInternal("dct_cases.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	var doc struct {
		N     int `json:"n"`
		Cases map[string]struct {
			Input  []float64 `json:"input"`
			Output []float64 `json:"output"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(data, &doc); err != nil {
		t.Fatalf("parse: %v", err)
	}
	for name, c := range doc.Cases {
		name, c := name, c
		t.Run(name, func(t *testing.T) {
			got := dct.DCT1D(c.Input)
			if len(got) != doc.N {
				t.Fatalf("length: got %d, want %d", len(got), doc.N)
			}
			for k := range got {
				if math.Abs(got[k]-c.Output[k]) > tol {
					t.Fatalf("%s k=%d: got %g, want %g", name, k, got[k], c.Output[k])
				}
			}
		})
	}
}

func TestArangeFirstOutputIs992(t *testing.T) {
	x := make([]float64, 32)
	for i := range x {
		x[i] = float64(i)
	}
	y := dct.DCT1D(x)
	if math.Abs(y[0]-992.0) > 1e-9 {
		t.Fatalf("y[0]: got %g, want 992.0", y[0])
	}
}

File: hash/java/src/test/java/io/rosetta/imagehash/group1/DctTest.java
package io.rosetta.imagehash.group1;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.rosetta.imagehash.internal.Dct;
import io.rosetta.imagehash.testkit.SpecPath;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.util.Iterator;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class DctTest {
    private static final double TOL = 1e-9;

    @Test
    void oneDimMatchesScipy() throws IOException {
        JsonNode data = new ObjectMapper().readTree(SpecPath.DCT_CASES.toFile());
        int n = data.get("n").asInt();
        JsonNode cases = data.get("cases");
        for (Iterator<Map.Entry<String, JsonNode>> it = cases.fields(); it.hasNext(); ) {
            Map.Entry<String, JsonNode> e = it.next();
            String name = e.getKey();
            JsonNode caseNode = e.getValue();
            double[] input = new double[n];
            for (int i = 0; i < n; i++) input[i] = caseNode.get("input").get(i).asDouble();
            double[] expected = new double[n];
            for (int i = 0; i < n; i++) expected[i] = caseNode.get("output").get(i).asDouble();
            double[] actual = Dct.dct1d(input);
            assertEquals(n, actual.length, name + ": length");
            for (int k = 0; k < n; k++) {
                int kk = k;
                assertTrue(Math.abs(actual[kk] - expected[kk]) < TOL,
                        () -> name + ": k=" + kk + " expected " + expected[kk] + ", got " + actual[kk] + ", diff " + (actual[kk] - expected[kk]));
            }
        }
    }

    @Test
    void arangeFirstOutputIs992() {
        double[] x = new double[32];
        for (int i = 0; i < 32; i++) x[i] = i;
        double[] y = Dct.dct1d(x);
        // y[0] = 2 * sum(x[n] * cos(0)) = 2 * sum(0..31) = 2 * 496 = 992
        assertEquals(992.0, y[0], 1e-9);
    }
}
